File: java/leet_code/1992.java
import java.util.*;

/**
 * This class provides a solution for finding all farmlands in a given 2D grid.
 * A farmland is defined as a group of '1's in the grid. Each group of '1's is
 * surrounded by '0's and forms a rectangle. The solution is to return the
 * coordinates of the top-left and bottom-right corners of each farmland.
 */
class Solution {

    /**
     * Finds all farmlands in the given 2D grid.
     *
     * @param land The 2D grid representing the land. '1's represent farmland and '0's represent non-farmland.
     * @return A 2D array where each element is an array of four integers representing the coordinates of the top-left and bottom-right corners of a farmland.
     */
    public int[][] findFarmland(int[][] land) {
        List<int[]> result = new ArrayList<>();

        int m = land.length;
        int n = land[0].length;

        // Iterate over the entire grid
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                // If the current cell is farmland ('1'), find its coordinates
                if (land[i][j] == 1) {
                    int[] coordinates = findFarmlandCoordinates(land, i, j);
                    result.add(coordinates);
                }
            }
        }

        // Convert the result to a 2D array
        return result.toArray(new int[result.size()][]);
    }

    /**
     * Finds the coordinates of the farmland that includes the given cell.
     *
     * @param land The 2D grid representing the land.
     * @param row The row index of the given cell.
     * @param col The column index of the given cell.
     * @return An array of four integers representing the coordinates of the top-left and bottom-right corners of the farmland.
     */
    private int[] findFarmlandCoordinates(int[][] land, int row, int col) {
        int[] coordinates = new int[4];
        coordinates[0] = row;
        coordinates[1] = col;

        int m = land.length;
        int n = land[0].length;

        int r = row;
        int c = col;

        // Find the bottom-right corner of the farmland group
        while (r < m && land[r][col] == 1) r++;
        while (c < n && land[row][c] == 1) c++;
        coordinates[2] = r - 1;
        coordinates[3] = c - 1;

        // Mark all cells in the farmland group as visited ('0')
        for (int i = row; i < r; i++) {
            for (int j = col; j < c; j++) {
                land[i][j] = 0;
            }
        }

        return coordinates;
    }
}

File: csharp/leet_code/1992.cs
/// <summary>
/// This class provides a solution for the problem of finding farmland in a 2D grid map.
/// A farmland is a group of '1's (land) connected horizontally and vertically.
/// </summary>
public class Solution
{
    /// <summary>
    /// This method finds all the farmlands in the given grid.
    /// </summary>
    /// <param name="land">A 2D grid map of '1's (land) and '0's (water).</param>
    /// <returns>An array of farmlands where each farmland is represented by an array of four integers [rowStart, colStart, rowEnd, colEnd].</returns>
    public int[][] FindFarmland(int[][] land)
    {
        int m = land.Length; // The number of rows in the grid.
        int n = land[0].Length; // The number of columns in the grid.
        bool[,] visited = new bool[m, n]; // A 2D array to keep track of visited cells.
        var result = new List<int[]>(); // A list to store the farmlands.

        // Iterate over each cell in the grid.
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                // If the cell is land and has not been visited, find the coordinates of the farmland and add it to the result.
                if (land[i][j] == 1 && !visited[i, j])
                {
                    int[] coordinates = FindCoordinates(land, i, j, visited);
                    result.Add(coordinates);
                }
            }
        }
        return result.ToArray(); // Return the farmlands.
    }

    /// <summary>
    /// This method finds the coordinates of a farmland in the grid.
    /// </summary>
    /// <param name="land">A 2D grid map of '1's (land) and '0's (water).</param>
    /// <param name="y">The row index of the current cell.</param>
    /// <param name="x">The column index of the current cell.</param>
    /// <param name="visited">A 2D array to keep track of visited cells.</param>
    /// <returns>An array of four integers [rowStart, colStart, rowEnd, colEnd] representing the farmland.</returns>
    private int[] FindCoordinates(int[][] land, int y, int x, bool[,] visited)
    {
        int yEnd = y; // The row index of the end of the farmland.
        int xEnd = x; // The column index of the end of the farmland.

        // Find the row index of the end of the farmland.
        while (yEnd < land.Length - 1 && land[yEnd + 1][x] == 1)
        {
            ++yEnd;
        }

        // Find the column index of the end of the farmland.
        while (xEnd < land[0].Length - 1 && land[y][xEnd + 1] == 1)
        {
            ++xEnd;
        }

        // Mark all the cells of the farmland as visited.
        for (int i = y; i <= yEnd; i++)
        {
            for (int j = x; j <= xEnd; j++)
            {
                visited[i, j] = true;
            }
        }

        // Return the coordinates of the farmland.
        return new int[] { y, x, yEnd, xEnd };
    }
}
